[quant][graphmode][fix] Make inserted child module names unique

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #27403 [quant][graphmode] Minor fix and turn off fold_convbn
* #27348 [quant][graphmode] Add e2e test for conv+bn
* #27347 [quant][graphmode][fix] Don't permute in quantized::conv2d pattern
* #27346 [quant][graphmode][fix] insert_prepack_unpack for conv
* #27269 [refactor] Remove `test_quantizer.py` and reuse one of its test in `test_quantization.py`
* #27239 [refactor] Factor out insertPrepackUnpackForLinear
* #27238 [refactor] InsertObservers
* **#27237 [quant][graphmode][fix] Make inserted child module names unique**

Summary:
Making inserted observer module and wrapper module names unique

Test Plan:
test_jit.py

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

Differential Revision: [D18182917](https://our.internmc.facebook.com/intern/diff/D18182917)

jerryzh168 committed:
[quant][graphmode][fix] Make inserted child module names unique

Summary:
Making inserted observer module and wrapper module names unique

Test Plan:
test_jit.py

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fix] Make inserted child module names unique"


Summary:
Making inserted observer module and wrapper module names unique

Test Plan:
test_jit.py

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fix] Make inserted child module names unique"


Summary:
Making inserted observer module and wrapper module names unique

Test Plan:
test_jit.py

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fix] Make inserted child module names unique"


Summary:
Making inserted observer module and wrapper module names unique

Test Plan:
test_jit.py

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fix] Make inserted child module names unique"


Summary:
Making inserted observer module and wrapper module names unique

Test Plan:
test_jit.py

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fix] Make inserted child module names unique"


Summary:
Making inserted observer module and wrapper module names unique

Test Plan:
test_jit.py

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fix] Make inserted child module names unique"


Summary:
Making inserted observer module and wrapper module names unique

Test Plan:
test_jit.py

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fix] Make inserted child module names unique"


Summary:
Making inserted observer module and wrapper module names unique

Test Plan:
test_jit.py

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fix] Make inserted child module names unique"


Summary:
Making inserted observer module and wrapper module names unique

Test Plan:
test_jit.py

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

ZolotukhinM commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -164,12 +179,15 @@ Node* InsertObserversHelper::insertObserverFor(
   }
-  std::string observer_name = "observer_for_" + v->debugName();
   script::Module observer = observer_module.clone();
+  std::string observer_name = "_observer_" + std::to_string(uid_.getUniqueId());
```
Why can't we just have `size_t uid = 0;` here and `uid++;` in the loop? What's the benefit of having a common enumeration across different uses? The check `module.find_module(observer_name)` is sufficient for correctness - if it passes then there are no issues even if this `uid` overlaps with `uid` used in another module.

jerryzh168 replied:
you mean always start from 0? then we are doing N^2 checks?

ZolotukhinM replied:
Frankly I think even N^2 won't be a problem here because number of observers won't be that big while readability of the code would improve by not introducing a new class. Also, you could move `size_t uid = 0;` to the beginning of this function and thus make sure that we're not rechecking the same `uid` more than once per method of the module - in real life I think it would mean that we're almost never rechecking them, so you get the O(N) complexity.

jerryzh168 replied:
I'm not sure I follow, this function is adding observer for one Value, how is moving `uid` to the beginning of this function prevent N^2 checks? Are you saying moving this outside of this function?

jerryzh168 replied:
I think we can remove the class by changing UniqueId to `int`

jerryzh168 committed:
Update on "[quant][graphmode][fix] Make inserted child module names unique"


Summary:
Making inserted observer module and wrapper module names unique

Test Plan:
test_jit.py

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fix] Make inserted child module names unique"


Summary:
Making inserted observer module and wrapper module names unique

Test Plan:
test_jit.py

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fix] Make inserted child module names unique"


Summary:
Making inserted observer module and wrapper module names unique

Test Plan:
test_jit.py

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fix] Make inserted child module names unique"


Summary:
Making inserted observer module and wrapper module names unique

Test Plan:
test_jit.py

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fix] Make inserted child module names unique"


Summary:
Making inserted observer module and wrapper module names unique

Test Plan:
test_jit.py

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fix] Make inserted child module names unique"


Summary:
Making inserted observer module and wrapper module names unique

Test Plan:
test_jit.py

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fix] Make inserted child module names unique"


Summary:
Making inserted observer module and wrapper module names unique

Test Plan:
test_jit.py

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fix] Make inserted child module names unique"


Summary:
Making inserted observer module and wrapper module names unique

Test Plan:
test_jit.py

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fix] Make inserted child module names unique"


Summary:
Making inserted observer module and wrapper module names unique

Test Plan:
test_jit.py

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fix] Make inserted child module names unique"


Summary:
Making inserted observer module and wrapper module names unique

Test Plan:
test_jit.py

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fix] Make inserted child module names unique"


Summary:
Making inserted observer module and wrapper module names unique

Test Plan:
test_jit.py

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fix] Make inserted child module names unique"


Summary:
Making inserted observer module and wrapper module names unique

Test Plan:
test_jit.py

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fix] Make inserted child module names unique"


Summary:
Making inserted observer module and wrapper module names unique

Test Plan:
test_jit.py

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fix] Make inserted child module names unique"


Summary:
Making inserted observer module and wrapper module names unique

Test Plan:
test_jit.py

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

ZolotukhinM commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -961,7 +966,8 @@ void FoldPrepackedWeightIntoModule(
     const script::Module& linear_params_module,
-    const script::Module& conv_params_module) {
+    const script::Module& conv_params_module,
+    int& uid) {
```
Can we just start from 0 in each module? That would avoid the extra argument here (note that to understand what this argument means one needs to know very deep details of the implementation of this function - that's why I'd prefer to avoid adding it).

ZolotukhinM commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -145,6 +145,8 @@ class InsertObserversHelper {
   // will be propagated through the function call hierarchy
   std::unordered_set<Value*> bias_values_;
+  // Unique id generator for observer module
+  int uid_;
```
Should it be initialized with 0?

jerryzh168 committed:
Update on "[quant][graphmode][fix] Make inserted child module names unique"


Summary:
Making inserted observer module and wrapper module names unique

Test Plan:
test_jit.py

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fix] Make inserted child module names unique"


Summary:
Making inserted observer module and wrapper module names unique

Test Plan:
test_jit.py

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

ZolotukhinM commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -145,6 +145,8 @@ class InsertObserversHelper {
   // Values that are the output of GetAttr[name="bias"] and they
   // will be propagated through the function call hierarchy
   std::unordered_set<Value*> bias_values_;
+  // Unique id generator for observer module
```
Nit: the comment might cover why we do this and when this is used, why we put it here and not right before the loop. Basically, cover more of "why" than of "what".

ZolotukhinM reviewed (approved):
Looks good!

jerryzh168 committed:
Update on "[quant][graphmode][fix] Make inserted child module names unique"


Summary:
Making inserted observer module and wrapper module names unique

Test Plan:
test_jit.py

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

Differential Revision: [D18182917](https://our.internmc.facebook.com/intern/diff/D18182917)

[ghstack-poisoned]
